### DISCUSSION TITLE
PPO 纯仿真流水线设计方案

### DISCUSSION THREAD

**Discussion by xieyucheng123** [link-seek/so101-sim-pipeline#2] [Ideas]
## 背景

当前 ACT/SmolVLA 训练后评估 success_rate=0%（P1 Sim-to-Real Gap + P2 训练不足）。SO101-Nexus 官方已有 seed-validated PPO baseline（`ppo_warp.py`），WarpPickLift-v1 上 5 seed 平均 success 0.973。PPO 仅用于纯仿真验证，不走 sim2real。

## 核心前提

| 项目 | 状态 |
|------|------|
| PPO 脚本 | SO101-Nexus `examples/ppo_warp.py`，CleanRL 风格 MLP actor-critic，GPU 并行 Warp 环境 |
| V100 + Warp | **可用** — PPO 用 `so101_nexus[warp]` 的 GPU 并行仿真（step+reward），不触发渲染路径的 `WarpCodegenError` |
| V100 + CUDA | 必须用 `cu126`（含 sm_70），当前 Dockerfile.train 已用 `cu126` |
| 评估 | PPO 自带 `evaluate_mujoco()` 在 MuJoCo 后端做确定性评估 + 视频渲染 |
| BC | Behavior Cloning（行为克隆），用遥操作 demo 预训练再 PPO fine-tune，解决 seed 脆弱性。本次先不做 BC+PPO，后续按需追加 |

## 官方 PPO baseline 结果（参考）

| 方式 | 脚本 | WarpPickLift-v1 success | 耗时 (RTX 5090) |
|------|------|--------------------------|-----------------|
| 纯 PPO | `ppo_warp.py` | 0.973 mean (5 seed) | 24.5 min |
| BC + PPO | `bc_ppo_warp.py` | 0.983 (seed 5 救回) | ~30 min |

## 交付物清单（5 个文件）

| # | 文件 | 说明 |
|---|------|------|
| 1 | `docker/Dockerfile.ppo` | PPO 镜像：`nvidia/cuda:12.6.3-runtime-ubuntu22.04` + `torch cu126` + `so101_nexus[warp,train]` + `imageio[ffmpeg]` + `opencv-python` |
| 2 | `scripts/train_ppo.py` | 从官方 `ppo_warp.py` 复制核心代码，适配流水线：argparse CLI、checkpoint → `/data/checkpoints/ppo/`、结果 JSON 输出 |
| 3 | `scripts/eval_ppo.py` | 加载 `best_agent.pt`，MuJoCo 后端 50 episodes 评估，输出 `eval_result.json` + `eval_video.mp4` |
| 4 | `.github/workflows/ppo-pipeline.yml` | start-ecs → ppo-train → ppo-eval → upload → stop-ecs |
| 5 | `.github/workflows/docker-build.yml` | 追加 `build-ppo` job |

## Dockerfile.ppo 设计

```dockerfile
FROM nvidia/cuda:12.6.3-runtime-ubuntu22.04

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get install -y --no-install-recommends \
    software-properties-common git wget curl build-essential cmake \
    libegl1 libgles2 libgl1-mesa-glx libvulkan1 mesa-vulkan-drivers ffmpeg \
    && add-apt-repository ppa:deadsnakes/ppa \
    && apt-get update && apt-get install -y python3.12 python3.12-venv python3.12-dev \
    && rm -rf /var/lib/apt/lists/*

RUN python3.12 -m venv /opt/venv
ENV PATH=\"/opt/venv/bin:$PATH\"
RUN pip install --no-cache-dir --upgrade pip setuptools wheel

RUN pip install --no-cache-dir torch==2.11.0 torchvision torchaudio \
    --index-url https://download.pytorch.org/whl/cu126

RUN pip install --no-cache-dir \"so101_nexus[warp,train]\" tyro \"imageio[ffmpeg]\" opencv-python

ENV MUJOCO_GL=egl
ENV NVIDIA_DRIVER_CAPABILITIES=compute,utility,graphics
WORKDIR /workspace
COPY scripts/ /workspace/scripts/
CMD [\"python\", \"/workspace/scripts/train_ppo.py\"]
```

## ppo-pipeline.yml 设计

```yaml
name: PPO Pipeline

on:
  workflow_dispatch:
    inputs:
      env_id:
        description: \"Warp env id\"
        default: \"WarpPickLift-v1\"
      total_timesteps:
        description: \"Training timesteps\"
        default: \"30000000\"
      num_envs:
        description: \"Parallel envs\"
        default: \"1024\"
      seed:
        description: \"Random seed\"
        default: \"1\"

env:
  PPO_IMAGE: swr.cn-north-4.myhuaweicloud.com/link-seek/so101-ppo:latest
  ECS_SERVER_ID: 7f39cb83-1a5c-4792-b65e-e578d7ddb88d
  ECS_REGION: cn-north-4
  OBS_BUCKET: so101-sim-pipeline

jobs:
  start-ecs:
    # 复用现有 ECS 启动逻辑 (~2min)

  ppo-pipeline:
    needs: start-ecs
    runs-on: [self-hosted, Linux, X64, V100]
    timeout-minutes: 120
    steps:
      - Login to SWR + Pull so101-ppo:latest

      - name: PPO Train          # ~40-50min on V100
        docker run --gpus all --shm-size=8g \
          -v /data/checkpoints:/data/checkpoints \
          -v /data/ppo:/data/ppo \
          so101-ppo:latest \
          python /workspace/scripts/train_ppo.py \
            --env-id WarpPickLift-v1 \
            --total-timesteps 30000000 \
            --num-envs 1024 --seed 1 \
            --save-dir /data/checkpoints/ppo \
            --result-path /data/ppo/results/train_result.json

      - name: PPO Eval           # ~8-10min, 50 episodes
        docker run --gpus all \
          -v /data/checkpoints:/data/checkpoints \
          -v /data/ppo:/data/ppo \
          so101-ppo:latest \
          python /workspace/scripts/eval_ppo.py \
            --checkpoint /data/checkpoints/ppo/best_agent.pt \
            --env-id WarpPickLift-v1 \
            --num-episodes 50 \
            --output /data/ppo/results

      - name: Upload to OBS + artifacts

  stop-ecs:
    # 复用现有 ECS 停止逻辑
```

## train_ppo.py 设计

- 从 `johnsutor/so101-nexus` 的 `examples/ppo_warp.py` 复制核心代码（Agent、ObsNormalizer、RewardScaler、train 函数）
- 原因：`examples/` 不是 importable 包，流水线需要自定义 CLI（argparse）和结果输出
- CLI 参数：`--env-id`, `--total-timesteps`, `--num-envs`, `--seed`, `--save-dir`, `--result-path`
- 训练完成后写 `train_result.json`：`{success_rate, best_success, total_timesteps, env_id, seed, sps, elapsed_s}`
- Checkpoint 保存 `best_agent.pt`（含 model state_dict + obs_norm stats）

## eval_ppo.py 设计

- 加载 `best_agent.pt`（含 model + obs_mean/obs_var）
- 在 MuJoCo 后端跑 50 个 episode（确定性，seed+1000+ep）
- 输出 `eval_result.json`：`{success_rate, avg_reward, avg_steps, num_episodes}`
- 渲染第一个 episode 为 `eval_video.mp4`

## PPO 超参数（WarpPickLift-v1 官方调优值）

| 参数 | 值 | 说明 |
|------|-----|------|
| `num_envs` | 1024 | GPU 并行环境数 |
| `num_steps` | 16 | 每次更新 rollout 长度 |
| `total_timesteps` | 30,000,000 | 总训练步数 |
| `learning_rate` | 3e-4 | 学习率（线性退火） |
| `gamma` | 0.99 | 折扣因子 |
| `gae_lambda` | 0.95 | GAE 参数 |
| `num_minibatches` | 32 | 小批量数 |
| `update_epochs` | 10 | 每次更新 epoch 数 |
| `clip_coef` | 0.2 | PPO clip 参数 |
| `ent_coef` | 0.03 | 熵系数（warm-start） |
| `ent_coef_final` | 0.005 | 熵系数终值（nonzero floor） |
| `vf_coef` | 0.5 | value loss 系数 |
| `max_grad_norm` | 0.5 | 梯度裁剪 |
| `hidden_dim` | 256 | MLP 隐藏层维度 |
| `episode_length` | 512 | 固定 horizon（避免 reach-farming） |
| `control_mode` | pd_joint_delta_pos | 控制模式 |

## V100 性能预估

| 指标 | RTX 5090 (参考) | V100-PCIE-32GB (预估) |
|------|-----------------|----------------------|
| 30M steps 耗时 | 24.5 min | ~40-50 min |
| SPS | ~20k | ~10-12k |
| 最终 success_rate | 0.973 | 应接近（算法不依赖 GPU 架构） |
| 总流水线耗时 | - | ~50-60 min |

## 执行顺序

1. 创建 `docker/Dockerfile.ppo`
2. 创建 `scripts/train_ppo.py`（从官方 ppo_warp.py 复制 + 适配）
3. 创建 `scripts/eval_ppo.py`
4. 创建 `.github/workflows/ppo-pipeline.yml`
5. 修改 `.github/workflows/docker-build.yml` 追加 build-ppo job
6. 推送代码 → 触发 docker-build → 构建 so101-ppo:latest → 推到 SWR
7. 手动触发 `ppo-pipeline.yml` → 验证端到端

## 与现有流水线的关系

| 流水线 | 训练方式 | 评估方式 | 目标 |
|--------|----------|----------|------|
| `vla-pipeline.yml` | SmolVLA IL | MuJoCo replay | sim2real（当前 0% success） |
| `warp-pipeline.yml` | ACT IL + Warp 数据采集 | MuJoCo eval | sim2real（当前 0% success） |
| `ppo-pipeline.yml` (新增) | PPO RL | MuJoCo eval (50 ep) | **纯仿真验证** |

PPO 流水线独立运行，不影响现有 VLA/Warp 流水线。PPO 是 MLP 策略（非视觉），不用于真机部署，仅验证 Warp 环境可学性和建立 RL baseline。

**Comment by xieyucheng123** [comment]
## PPO 流水线端到端验证结果

### 流水线运行

| 项目 | 值 |
|------|-----|
| Run | https://github.com/link-seek/so101-sim-pipeline/actions/runs/31767540540 |
| 触发方式 | workflow_dispatch |
| 参数 | WarpPickLift-v1, 30M steps, 1024 envs, seed=1 |
| 总耗时 | ~89 min (start-ecs 2min + train 86min + eval 14s + stop-ecs 1min) |
| 结论 | **SUCCESS** |

### 训练结果

| 指标 | 值 |
|------|-----|
| env_id | WarpPickLift-v1 |
| total_timesteps | 30,000,000 |
| best_success | **0.995** |
| final_success | 0.965 |
| elapsed | 5186s (~86 min) |
| SPS | 5784 steps/s |

### 评估结果（MuJoCo 后端，50 episodes 确定性评估）

| 指标 | 值 |
|------|-----|
| **success_rate** | **1.000** (50/50 全部成功) |
| avg_reward | 1.440 |
| avg_steps | 50.0 |
| elapsed | 13.8s |

### 与官方 baseline 对比

| 环境 | GPU | 30M steps 耗时 | SPS | best_success | eval success_rate |
|------|-----|----------------|-----|-------------|-------------------|
| 官方 RTX 5090 | sm_90 | 24.5 min | ~20k | 0.973 (5 seed mean) | - |
| **本项目 V100** | sm_70 | **86 min** | **5784** | **0.995** | **1.000** |

V100 比 RTX 5090 慢约 3.5x（5784 vs ~20k SPS），但最终结果更好（0.995 vs 0.973），可能因为 seed 运气。

### 修复记录

首次运行遇到 mujoco_warp sensor 模块 NVRTC 编译错误（undefined _magnetometer_0, _cam_projection_0 等符号），根因是 so101_nexus 0.5.1 限制 mujoco-warp<3.10，而 3.9.x 的 sensor.py 有代码生成 bug。

修复：在 Dockerfile.ppo 中追加 pip install mujoco-warp>=3.10.0.1,<3.12，升级到包含 sensor 修复（PR #1446, #1447）的版本。

### 交付物

| 文件 | commit |
|------|--------|
| docker/Dockerfile.ppo | 2cfb130, 8afd173 |
| scripts/train_ppo.py | 2cfb130 |
| scripts/eval_ppo.py | 2cfb130 |
| .github/workflows/ppo-pipeline.yml | 2cfb130 |
| .github/workflows/docker-build.yml | 2cfb130 (追加 build-ppo job) |

### OBS 产物

- obs://so101-sim-pipeline/ppo/train_result.json
- obs://so101-sim-pipeline/ppo/eval_result.json
- obs://so101-sim-pipeline/ppo/eval_video.mp4
- obs://so101-sim-pipeline/ppo/checkpoints/best_agent.pt (591KB)

### 结论

PPO 纯仿真流水线端到端验证通过。WarpPickLift-v1 上 V100 训练 30M steps 后，MuJoCo 后端 50 episode 确定性评估 success_rate=**100%**，证明：
1. Warp GPU 并行环境在 V100 (sm_70) 上可正常工作（仅需 mujoco_warp>=3.10）
2. PPO CleanRL recipe 在 V100 上收敛，与官方 RTX 5090 结果一致
3. 流水线基础设施（ECS + Docker + SWR + OBS）端到端可用

**Comment by xieyucheng123** [comment]
## PPO v2 结果：lift_threshold=0.15（15cm 明显抬起）

### 背景

v1 的 lift_threshold=0.05（5cm）太低，策略学会微抬 5cm 即判定成功，视频里看不出抬起动作。提高到 0.15（15cm）后重新训练+评估。

### 流水线运行

| 项目 | 值 |
|------|-----|
| Run | https://github.com/link-seek/so101-sim-pipeline/actions/runs/31788453128 |
| 参数 | WarpPickLift-v1, 30M steps, lift_threshold=0.15 |
| 训练耗时 | 5112s (~85min) |
| SPS | 5867 |
| Commit | b09f762 (feat: add --lift-threshold CLI arg) |

### 训练结果

| 指标 | v1 (threshold=0.05) | v2 (threshold=0.15) |
|------|---------------------|---------------------|
| best_success | 0.995 | 0.9925 |
| final success_rate | 0.965 | 0.975 |
| iterations | 5186 | 1831 |
| episodes | 400 | 400 |

### 评估结果（MuJoCo, 50 episodes）

| 指标 | v1 (threshold=0.05) | v2 (threshold=0.15) |
|------|---------------------|---------------------|
| success_rate | 1.000 (50/50) | 0.980 (49/50) |
| avg_reward | 1.440 | 1.511 |
| avg_steps | 50.0 | 58.4 |

### 结论

- **success_rate 仅下降 2%**（1.000 → 0.980），说明策略能可靠抓取并抬起 15cm
- **avg_reward 提高**（1.440 → 1.511），因为更高 lift_threshold 给出更高 reward 信号
- **avg_steps 增加**（50.0 → 58.4），策略需要更多步骤完成更高抬起
- 唯一失败 episode (34) 跑满 512 步未成功，其余 49 个均成功
- 视频中物体应有明显抬起动作（15cm vs 之前 5cm）

### 交付物

- eval_video.mp4 / train_result.json / eval_result.json: obs://so101-sim-pipeline/ppo/
- checkpoint: obs://so101-sim-pipeline/ppo/checkpoints/best_agent.pt

### 下一步

PPO 仿真策略已验证可靠。接下来将执行 Discussion #4：复用 ataghof/so101nexus-cube500-binary 数据集训练 SmolVLA，验证 VLA 方案能否消除 P1 视觉域不匹配问题。

**Comment by xieyucheng123** [comment]
## 收尾（工作已完成）

PPO v2（lift_threshold=0.15，run `31788453128`）：MuJoCo 50eps 49/50 = **98%**，avg_reward 1.511，训练+评估+视频交付物齐。纯仿真验证目标达成。

本帖使命完成，RESOLVED 关闭。
